src/kubernetes/reflector.rs

```diff
@@ -0,0 +1,1037 @@
+//! Watch and cache the remote Kubernetes API resources.
+
+use super::{
+    resource_version, state,
+    watcher::{self, Watcher},
+};
+use crate::internal_events::kubernetes::reflector as internal_events;
+use futures::{
+    pin_mut,
+    stream::{Stream, StreamExt},
+};
+use k8s_openapi::{
+    apimachinery::pkg::apis::meta::v1::{ObjectMeta, WatchEvent},
+    Metadata, WatchOptional, WatchResponse,
+};
+use snafu::Snafu;
+use std::convert::Infallible;
+use std::time::Duration;
+use tokio::{select, time::delay_for};
+
+/// Watches remote Kubernetes resources and maintains a local representation of
+/// the remote state. "Reflects" the remote state locally.
+///
+/// Does not expose evented API, but keeps track of the resource versions and
+/// will automatically resume on desync.
+pub struct Reflector<W, S>
```

Comment:
It seems like you might be able to contribute some of this to https://github.com/clux/kube-rs/blob/master/kube/src/runtime/reflector.rs, they're upgrading futures right now?

Reply:
I did consider it, but I'm focusing on making things work for us first. Technically we don't even have to contribute directly to `kube` crate - our reflector code is properly abstracted and should work with any watcher implementation (as long as there's an adapter). But that's not the focus atm. Good idea though, I'd be happy to publish the parts of this code as independent pieces.

Reply:
Thinking ahead, it might be worthwhile for us to attempt to push down some of this to them and help them refine it a bit, it'd relieve some of the maintenance burden off us and help the community.

Reply:
I wouldn't mind that either. Ideally, we'd want to port most of the code out to somewhere. Not sure if `kube` crate is a good candidate for that, since they have some specifics that prevented us from using it in the first place. We could work with them to fix the issue, and then we can contribute our parts. But that's not the only option, and I'd handle it differently in the current layout.
Let's deal with it after the merge though, otherwise it can drag for an unlimited amount of time :smile: